examples/Custom_Hooks/probability_of_improvement.py

```
@@ -0,0 +1,219 @@
+    pi = botorch_acqf(comp_rep_tensor)
+    pi = pi + 1e-10 * torch.randn(pi.shape)
+    pi_list.append(pi)
+
+
+# Additionally, we define a function that plots the `poi_list` after all recommend  iterations:
+
+
+def plot_pi(
+    pi_list: list[torch.Tensor],
+    ax: Axes,
+    base_name: str,
+    path: Path = Path("."),
+) -> Figure:
+    """Plot the probability of improvement in 3D."""
+    cmap = plt.get_cmap("viridis")
+
+    pi_max = max([torch.max(p).item() for p in pi_list])
+
+    # Plot each PI tensor separately
+    for i, p in enumerate(pi_list):
+        pi_np = p.detach().numpy() if p.requires_grad else p.numpy()
+        x = np.linspace(0, pi_max, 500)
+        kde = gaussian_kde(pi_np)
+        y = kde(x)
+        z = np.full_like(y, i)
+
+        # Fill under the curve
+        verts = []
+        verts.append([(x[i], 0.0), *zip(x, y), (x[-1], 0.0)])
+        color = cmap(float(i) / len(pi_list))
+        poly = PolyCollection(verts, color=color, alpha=0.9)
+        ax.add_collection3d(poly, zs=i, zdir="x")
+
+        ax.plot(x, y, z, zdir="x", color=color)
+
+    # Set viewing angle
+    ax.view_init(elev=20, azim=30)
+
+    # Reduce space between the iterations
+    ax.set_box_aspect([0.7, 1, 1])
+
+    # Set the y-axis limit based on the maximal PI
+    ax.set_ylim(0, pi_max)
+
+    # Set x-axis ticks to have the correct iteration number
+    ax.set_xticks(np.arange(0, len(pi_list), 1))
+    ax.set_xticklabels([i for i in range(1, len(pi_list) + 1)])
+
+    ax.set_ylabel("PI")
+    ax.set_xlabel("Iteration")
+    ax.set_zlabel("Density")
+    ax.set_title("Investigate the PI")
+
+    output_path = Path(path, base_name)
+    # mypy thinks that ax.figure might become None, hence the explicit ignore
+    if isinstance(ax.figure, Figure):
+        ax.figure.savefig(
+            output_path,
+            format="svg",
+            transparent=True,
+        )
+    else:
+        warnings.warn("Plots could not be saved.")
```

id rather let this fail or is failure here expected often?

---

Honestly, I am not sure. @AVHopp what do you think?

---

Resolving this, since this is not part of `create_pi_plot()` anymore.